src/provider/provider_tracking.c

```diff
@@ -119,6 +121,164 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
     return ret;
 }
 
+static umf_result_t trackingAllocationSplit(void *hProvider, void *ptr,
+                                            size_t totalSize,
+                                            size_t firstSize) {
+    umf_result_t ret = UMF_RESULT_ERROR_UNKNOWN;
+    umf_tracking_memory_provider_t *provider =
+        (umf_tracking_memory_provider_t *)hProvider;
+
+    tracker_value_t *splitValue =
+        (tracker_value_t *)malloc(sizeof(tracker_value_t));
+    if (!splitValue) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    splitValue->pool = provider->pool;
+    splitValue->size = firstSize;
+
+    int r = util_mutex_lock(provider->hTracker->splitMergeMutex);
+    if (r) {
+        goto err_lock;
+    }
+
+    tracker_value_t *value =
+        (tracker_value_t *)critnib_get(provider->hTracker->map, (uintptr_t)ptr);
+    if (!value) {
+        fprintf(stderr, "tracking split: no such value\n");
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err;
+    }
+    if (value->size != totalSize) {
+        fprintf(stderr, "tracking split: %zu != %zu\n", value->size, totalSize);
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err;
+    }
+
+    ret = umfMemoryProviderAllocationSplit(provider->hUpstream, ptr, totalSize,
+                                           firstSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(stderr,
+                "tracking split: umfMemoryProviderAllocationSplit failed\n");
+        goto err;
+    }
+
+    void *highPtr = (void *)(((uintptr_t)ptr) + firstSize);
+    size_t secondSize = totalSize - firstSize;
+
+    // We'll have a duplicate entry for the range [highPtr, highValue->size] but this is fine,
+    // the value is the same anyway and we forbid removing that range concurrently
+    ret = umfMemoryTrackerAdd(provider->hTracker, provider->pool, highPtr,
+                              secondSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "tracking split: umfMemoryTrackerAdd failed\n");
+        // TODO: what now? should we rollback the split? This can only happen due to ENOMEM
```

Comment:
yes, preallocation makes sense - should be doable once we have the base allocator.